python/lance/_lib.pyx

```diff
@@ -273,28 +276,43 @@ cdef class FileSystemDataset(Dataset):
         c_version = GetResultValue(self.lance_dataset.latest_version())
         return _dataset_version_to_json(c_version)
 
-    def append_column(self, field: Field, func: Callable[[pyarrow.Table], pyarrow.Array]) -> FileSystemDataset:
+    def _append_column_expr(self, field: Field, expression: Expression) -> FileSystemDataset:
+        cdef CExpression c_expression = expression.unwrap()
+        cdef shared_ptr[CField] c_field = pyarrow_unwrap_field(field)
+        cdef shared_ptr[CLanceDataset] dataset = GetResultValue(self.lance_dataset.AddColumn(c_field, c_expression))
+        return FileSystemDataset.wrap(static_pointer_cast[CDataset, CLanceDataset](dataset))
+
+    def append_column(
+            self,
+            field: Field,
+            value: Union[Callable[[pyarrow.Table], pyarrow.Array], Expression],
```

Comment:
Scalar is not listed as a supported type?

Reply:
Scalar is not. `Expression.literal()` is.